Fixes to posting problems in the tree, steiner and dag globals, and fix to the incorrect unsatisfiability with the connected constraint

Samclou opened the pull request:
Fix #140

For the problem with the posting of the tree constraint, the root parameter has simply been removed from the chuffed_tree predicate in the relevant mzn. To maintain consistency with the Minizinc decomposition, a constraint verifying that the root is in the sub-graph is added.

For the steiner constraint, the fix is simply to initialize the weight vector when posting.

Finally, for dag, the first fix is to make sure that the new edges and the new node are added in all the relevant vectors. For that, it suffice to update "nb_nodes" and "nb_edges" before the loops and adding 1 to the nodes added in "from" and "to" (to be consistent with the node indexes in these vectors). The rest of the modifications concerns the new edges added. By having them be "bv_true", all nodes given were also set to true. From my understanding, the reason an extra node is added is so that the DReachability propagator can have a fixed root without us having to choose it in the graph given. By treating the new edges as unfixed boolean variables and adding a cardinality constraint on them should guaranty the wanted behaviour.

Samclou committed:
Fixes to posting problems in the graph-related globals

Modification to fzn_tree_int.mzn to prevent an inconsistency in the number of parameters received in the tree poster.
Modification of the steiner_tree poster so that the weight vector is initialized correctly.
Modification of the dag poster so that the new edges and node are added correctly in the relevant vectors. Also modified the nature of the new edges, as to not make every node mandatory. A bool_linear constraint is posted to deal with these new edges.

Dekker1 commented:
Thank you for pointing out the problem, and providing the fix. I'm just returning from leave, so it might take a little while before I have time to have a in-depth look at this. Apologies for the delay.

Samclou committed:
Correct an inheritance issue between the tree and the connected propagators.

ConnectedPropagator is defined as a subclass of TreePropagator in which some methods are redefined to deactivate the propagations relative to cycles. These methods were not made virtual so ConnectedPropagator had the same behavior as its parent.
Making the methods virtual lets the deactivation work. The new redefinition of the unite method is simply to remove an assertion regarding cycles that is probably useful when debugging the TreePropagator but becomes troublesome in ConnectedPropagator.

Samclou commented:
It seems that I messed up my commits so I will explain the second one here too.

Fix #137

The Connected propagator is defined as a subclass of the Tree one where some methods regarding cycles are redefined to deactivate the Acyclicity part of the Tree propagator. Since those methods were not declared virtual, the redefinitions were not used and the Connected propagator was behaving like the Tree one. Declaring the relevant methods as virtual fixes that.

Dekker1 commented:
The changes have been rebased and merged onto develop
